openedx/openedx-platform
# Fix  scorm xblock x frame options

## TL;DR

This adds support for the [SCORM xblock](https://github.com/overhangio/openedx-scorm-xblock) at its latest version. Via that xblock, this allows to use SCORM blocks in courses.
This is done via a Content-Security-Policy (CSP) but only for specific endpoints.

## Description

To support SCORM Connect, we need to update [The SCORM xblock](https://github.com/overhangio/openedx-scorm-xblock) to the latest version (> 18) because that fixes a bug where the SCORM API does not provide learner_id.

When we initially tried to update this xblock, we learned that SCORM does not work on edx-platform with the default config of setting the `X-Frame-Options` response header to "DENY", because the SCORM xblock uses an iframe and that header disallows framing content.

To prevent clickjacking, either a `Content-Security-Policy` header with the `frame-ancestors` key must be in place, or `X-Frame-Options` must be set to "DENY" or "SAMEORIGIN".

`X-Frame-Options`: "SAMEORIGIN" is insufficient because we need to allow framing content by our Microfrontends, which have different-origin URLs.

So here we add a Content-Security-Policy header, but only to specific endpoints, because we don't want to make sweeping changes to edx-platform responses unless necessary.

To achieve this, we copied the Content-Security-Policy middleware from [edx-django-utils](https://github.com/openedx/edx-django-utils/blob/master/edx_django_utils/security/csp/middleware.py) and then modified the copy so that it allows adding different CSPs to specific endpoints via url regex matching.

## Supporting information

- [Internal Issue](https://2u-internal.atlassian.net/jira/software/c/projects/TNL/boards/744?selectedIssue=TNL-11641)
- [Internal Slack Convo about the content of the CSP header](https://twou.slack.com/archives/C02PVMPKB/p1719848915789029)


## Testing instructions

- SCORM Connect Test file: 
[edx-test-course-scorm-connect--scorm-test-pano--edXSCORMcourses--scorm-2004 (1).zip](https://github.com/user-attachments/files/16697496/edx-test-course-scorm-connect--scorm-test-pano--edXSCORMcourses--scorm-2004.1.zip)
- Sample default SCORM test file: 
[ContentPackagingSingleSCO_SCORM12.zip](https://github.com/user-attachments/files/16697548/ContentPackagingSingleSCO_SCORM12.zip)
- in devstack, do these steps each for CMS and LMS in this order:
- `make dev.shell.lms` (or cms)
- `make requirements`
- `pip install openedx-scorm-xblock` -> check that this is version > 18
- exit shell
- `make dev.restart-container.lms` (or cms)
- after this is done, go to studio, go to a course and to advanced settings, and in the "advanced modules" field right in the top, add "scorm" to the array.
- Go to a Unit and create an advanced block -> select SCORM (option should be visible if all above steps were performed correctly; if not visible you can retry those steps or reach out to @jesperhodge )
- edit the block and upload the SCORM Connect Test file
- repeat this for the Sample default SCORM test file
- now both blocks should display content
- click on Preview
- you should see both block contents
- open in LMS
- you should also see both block contents

## Deadline

We'd like to deploy this to stage if possible Monday morning.

## Other information

There's plenty of background info about SCORM Connect and why we're trying to support it if needed, just reach out to @jesperhodge .
